### PR TITLE
add custom metrics support

### PR DESCRIPTION
**Pull request by tlhunter**
### What does this PR do?
- adds support for sending custom metrics via a new API

### Motivation
- when using dd-trace-js at a previous employer I was always confused as to why this wasn't included

### Plugin Checklist

- [x] Unit tests.
- [X] TypeScript [definitions][1].
- [x] TypeScript [tests][2].
- [x] API [documentation][3].
- [ ] CircleCI [jobs/workflows][4].
- [X] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/documentation/blob/master/content/en/tracing/trace_collection/library_config/nodejs.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/index.js

### Additional Notes
- ~~still need to refactor runtime metrics to add histogram support~~
  - may add this in a future iteration
  - apparently distributions are better for many reasons
- some docs are added in https://github.com/DataDog/documentation/pull/19318
- other docs are in private repo for the [Metric Submission: DogStatsD](https://docs.datadoghq.com/metrics/custom_metrics/dogstatsd_metrics_submission) page

### CODE REVIEW

**Comment by tlhunter** [comment]
Rename `tracer.metrics` to `tracer.dogstatsd`.

**Comment by tlhunter** [comment]
Create an intermediary class that translates `{key:value}` pairs into `['key:value']`.

File: packages/dd-trace/src/dogstatsd.js [hunk 1]
```diff
@@ -153,7 +153,67 @@ class NoopDogStatsDClient {
...
+  }
+
+  flush () {
+    return this.dogstatsd.flush()
+  }
+
+  /**
+   * Exposing { tagName: 'tagValue' } to the end user
+   * These are translated into [ 'tagName:tagValue' ] for internal use
+   */
```
Comment (Qard) [inline packages/dd-trace/src/dogstatsd.js:201]: Any reason not to support array form externally too and just do the conversion if it's not _already_ an array?

File: index.d.ts [hunk 2]
```diff
@@ -642,6 +644,47 @@ export declare interface User {
...
 }
 
+export declare interface DogStatsD {
+  /**
+   * Increments a metric by the specified value, optionally specifying tags.
+   * @param {string} stat The dot-separated metric name.
+   * @param {number} value The amount to increment the stat by.
+   * @param {[tag:string]:string} tags Tags to pass along, such as `[ 'foo:bar' ]`. Values are combined with config.tags.
+   */
+  increment(stat: string, value?: number, tags?: { [tag: string]: string }): void
```
Comment (Qard) [inline index.d.ts:654]: Should the values also accept numbers?

**Review by Qard — COMMENTED** [review]
Generally LGTM, just a few questions.

File: packages/dd-trace/src/dogstatsd.js (hunk 1 above)
Comment (tlhunter) [reply packages/dd-trace/src/dogstatsd.js:201]: It's convenient for us but I can't imagine an end user ever wanting to provide data in that format. All other products also use key/value object pairs. We can always support arrays in the future if anyone requests it.

**Review by Qard — APPROVED** [review]

**Comment by tlhunter** [comment]
This is a fairly big new feature, one that I expect to go through a few iterations. For that reason I'm not going to add it to the final 2.x release.

**Pull request merged**